# Pharma Industry Drug Response Prediction

This notebook demonstrates a machine learning workflow to predict drug response in the pharma industry. We will use a Logistic Regression model for this  classification task. The steps include data loading, defining features and target, splitting data, training the model, and evaluating its performance.

In [7]:
#Load the data set
import pandas as pd
df = pd.read_csv("/content/Pharma_Industry.csv")
df

,Drug Dosage (mg),Systolic Blood Pressure (mmHg),Heart Rate (BPM),Liver Toxicity Index (U/L),Blood Glucose Level (mg/dL),Drug Response
0,-0.128538,0.303280,-1.881849,0.258286,-0.792011,1
1,-1.846188,2.865142,-0.929511,2.866786,-0.719447,1
2,-1.252393,-1.541613,0.363632,-0.325370,0.191314,0
3,1.992515,-1.142779,-0.766657,0.975286,-0.823355,1
4,0.377100,0.538410,-0.029263,1.896015,-0.960130,1
...,...,...,...,...,...,...
495,-1.553759,0.506938,0.817890,-0.661245,-1.345393,0
496,0.551476,1.714321,0.340589,1.712558,-0.496315,1
497,-1.072743,-0.221137,0.239247,1.112924,-0.285838,1
498,1.661259,0.245759,-0.457096,0.633355,0.657413,0


### Step 1: Define Features (X) and Target (y)

Before training any machine learning model, it's crucial to distinguish between your input features (the data points used for prediction) and your target variable (what you are trying to predict). In this case, 'Drug Response' is our target variable, and all other columns will be used as features.



In [8]:
from sklearn.model_selection import train_test_split

# Define features (X) and target (y)
X = df.drop('Drug Response', axis=1)  # All columns except 'Drug Response'
y = df['Drug Response']  # 'Drug Response' column is the target

print(f"Features (X) shape: {X.shape}")
print(f"Target (y) shape: {y.shape}")

Features (X) shape: (500, 5)
Target (y) shape: (500,)


### Step 2: Split Data into Training and Testing Sets

To ensure our model can generalize well to unseen data, we split our dataset into two parts:

*   **Training Set**: Used to train the machine learning model. The model learns patterns and relationships from this data.
*   **Testing Set**: Used to evaluate the model's performance on new, unseen data. This helps us understand how well the model will perform in real-world scenarios.

We typically use `train_test_split` from `sklearn.model_selection` to perform this split. A common split ratio is 80% for training and 20% for testing.

In [9]:
# Split data into training and testing sets (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

X_train shape: (400, 5)
X_test shape: (100, 5)
y_train shape: (400,)
y_test shape: (100,)


### Step 3: Train the Logistic Regression Model

Now that our data is prepared, we can initialize and train our Logistic Regression model. The `LogisticRegression` class from `sklearn.linear_model` is used for this purpose. The `fit()` method is called with the training features (`X_train`) and training target (`y_train`) to allow the model to learn the underlying relationships.

In [10]:
from sklearn.linear_model import LogisticRegression

# Create a Logistic Regression model
model = LogisticRegression(random_state=42)

# Train the model using the training data
model.fit(X_train, y_train)

print("Logistic Regression model trained successfully.")

Logistic Regression model trained successfully.


### Step 4: Evaluate the Model's Performance

After training, it's essential to evaluate how well our model performs on the test set. We make predictions on `X_test` and then compare these predictions (`y_pred`) to the actual target values (`y_test`). Common classification metrics used for evaluation include:

*   **Accuracy**: The proportion of correctly classified instances.
*   **Precision**: Of all instances predicted as positive, what proportion are actually positive.
*   **Recall**: Of all actual positive instances, what proportion did the model correctly identify.
*   **F1-Score**: The harmonic mean of Precision and Recall, providing a single metric that balances both.

In [11]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Make predictions on the test set
y_pred = model.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print(f"Model Accuracy: {accuracy:.4f}")
print(f"Model Precision: {precision:.4f}")
print(f"Model Recall: {recall:.4f}")
print(f"Model F1-Score: {f1:.4f}")

Model Accuracy: 0.7900
Model Precision: 0.8182
Model Recall: 0.8036
Model F1-Score: 0.8108


## Conclusion

This notebook successfully implemented a Logistic Regression model to predict drug response based on the provided pharmaceutical dataset. We covered the essential steps from data preparation to model evaluation. The model achieved a satisfactory performance with an accuracy of 0.79, indicating its potential utility in identifying factors influencing drug efficacy. Further analysis could involve exploring other machine learning algorithms, hyperparameter tuning, or feature engineering to potentially improve the model's predictive power.